torch_xla/csrc/aten_xla_type.cpp

```diff
@@ -1424,11 +1424,30 @@ at::Tensor XLANativeFunctions::index_add(const at::Tensor& self, int64_t dim,
 at::Tensor XLANativeFunctions::index_copy(const at::Tensor& self, int64_t dim,
                                           const at::Tensor& index,
                                           const at::Tensor& source) {
-  TORCH_LAZY_FN_COUNTER_TIMED_TRACING("xla::");
   XLATensorPtr self_tensor = bridge::GetXlaTensor(self);
-  return bridge::AtenFromXlaTensor(
-      tensor_methods::index_copy(self_tensor, dim, bridge::GetXlaTensor(index),
-                                 bridge::GetXlaTensor(source)));
+  absl::Span<const int64_t> self_tensor_dims =
+      self_tensor->shape().get().dimensions();
+
+  XLATensorPtr source_tensor = bridge::GetXlaTensor(source);
+  absl::Span<const int64_t> source_tensor_dims =
+      source_tensor->shape().get().dimensions();
+
+  XLATensorPtr index_tensor = bridge::GetXlaTensor(index);
+  absl::Span<const int64_t> index_tensor_dims =
+      index_tensor->shape().get().dimensions();
+
+  // As per
+  // https://pytorch.org/docs/stable/generated/torch.Tensor.index_copy_.html,
+  // the dim-th dimension of source tensor must have the same size as the length
+  // of index tensor.
+  XLA_CHECK(dim < source_tensor_dims.size() &&
+            source_tensor_dims[dim] == index_tensor_dims[0])
```

Comment:
For out of bound index issues, we need to verify the following test scenario is caught in tests.

```
x = torch.zeros(5, 3)
t = torch.tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9]], dtype=torch.float)
index = torch.tensor([0, 9, 2]) # the second index (i.e. 9) is out of bound as the first dim of x is 5.
x.index_copy_(0, index, t)
```